File: post24_async.js
function sleep(seconds) {
  return new Promise((resolve) => setTimeout(resolve, seconds / 1000));
}
async function goodCitizen(name) {
  console.log(`GOOD citizen ${name} starting`);
  for (let i = 0; i < 5; i++) {
    console.log(`GOOD citizen ${name}: `, i);
    // This 'await' yields control, allowing other tasks to run.
    await sleep(0.001);
  }
  console.log(`GOOD citizen ${name} ending`);
}
async function fakeAsync(x) {
  console.log("running fake async");
  var throwaway = 0;
  for (let i = 0; i < 1230000000; i++) {
    throwaway += i;
  }
  return x;
}
// Equivalent to Python's bad_citizen
async function badCitizen() {
  console.log("BAD citizen starting");
  for (let i = 0; i < 100; i++) {
    console.log("BAD citizen: ", i);
    await fakeAsync(i);
  }
  console.log("BAD citizen ending");
}
async function main() {
  const good1 = goodCitizen("John");
  const bad = badCitizen();
  const good2 = goodCitizen("Mary");
  await Promise.all([good1, good2, bad]);
}

main();

File: post24_async.py
import asyncio


async def good_citizen(name):
    print(f"GOOD citizen {name} starting")
    for i in range(5):
        print(f"GOOD citizen {name}: ", i)
        await asyncio.sleep(0.001)
    print(f"GOOD citizen {name} ending")


async def bad_citizen():
    print("BAD citizen starting")
    for i in range(5):
        print("BAD citizen: ", i)
        await fake_async(i)  # DOES NOT yield to event loop!!!


async def fake_async(x):
    print("running fake async")
    return x


async def main():
    async with asyncio.TaskGroup() as tg:
        tg.create_task(good_citizen("John"))
        tg.create_task(bad_citizen())
        # bad citizen will run to completion before good_citizen's loop can run
        tg.create_task(good_citizen("Mary"))


if __name__ == "__main__":
    asyncio.run(main())
